### PR TITLE
Phase 6: Diverse Hyperparameter Ensemble — Mixed Configs

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Our current 8-seed ensemble uses IDENTICAL configs with only seed diversity. The diversity comes solely from random initialization. Models with **different hyperparameters** make systematically different errors, providing HIGHER diversity for ensemble averaging. A 8-model "diverse ensemble" with mixed lr/T_max/wd could outperform an 8-model same-config ensemble.

Theory: The bias-variance-diversity decomposition (JMLR 2023) shows that ensemble improvement scales with DIVERSITY, not just model count. Different hyperparams create models that are good at different things.

## Instructions
No code changes needed. Train 8 models with different hyperparameter settings, all using the standard baseline command as the starting point.

Full baseline command:
```bash
python train.py --agent nezuko --wandb_group phase6/diverse-hparam-ensemble \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

| GPU | wandb_name | Variation | Flags that differ |
|-----|-----------|-----------|-------------------|
| 0 | nezuko/baseline-lr2e4-s42 | Standard baseline (control) | --lr 2e-4 --weight_decay 5e-5 --cosine_T_max 160 --seed 42 |
| 1 | nezuko/lr-1.5e4-s42 | Lower LR | --lr 1.5e-4 --weight_decay 5e-5 --cosine_T_max 160 --seed 42 |
| 2 | nezuko/lr-2.5e4-s42 | Higher LR | --lr 2.5e-4 --weight_decay 5e-5 --cosine_T_max 160 --seed 42 |
| 3 | nezuko/tmax-140-s42 | Faster decay | --lr 2e-4 --weight_decay 5e-5 --cosine_T_max 140 --seed 42 |
| 4 | nezuko/tmax-180-s42 | Slower decay | --lr 2e-4 --weight_decay 5e-5 --cosine_T_max 180 --seed 42 |
| 5 | nezuko/wd-1e5-s42 | Lower weight decay | --lr 2e-4 --weight_decay 1e-5 --cosine_T_max 160 --seed 42 |
| 6 | nezuko/wd-1e4-s42 | Higher weight decay | --lr 2e-4 --weight_decay 1e-4 --cosine_T_max 160 --seed 42 |
| 7 | nezuko/ema-0.9995-s42 | Higher EMA | --lr 2e-4 --weight_decay 5e-5 --cosine_T_max 160 --ema_decay 0.9995 --seed 42 |

ALL use seed 42 — the diversity comes from hyperparams, not seeds.

**What to report:**
1. Individual model surface MAE for each config
2. The 8-model diverse ensemble MAE (average all 8 predictions)
3. Compare diverse-8 ensemble vs the standard 8-seed ensemble (p_in=12.4, p_oodc=6.7, p_tan=29.4, p_re=5.8)
4. Compare diverse-8 vs the best 4-of-8 diverse subset

## Baseline
8-seed same-config ensemble: p_in=12.4, p_oodc=6.7, p_tan=29.4, p_re=5.8
Single-model mean: p_in=13.03±0.39, p_oodc=7.83±0.19, p_tan=30.29±0.47, p_re=6.45±0.05
W&B: wandb-applied-ai-team/senpai-v1

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Individual Model Performance (all seed 42, different hyperparams)

| GPU | Config | p_in | p_oodc | p_tan | p_re | val_loss | W&B Run ID |
|-----|--------|------|--------|-------|------|----------|------------|
| 0 | **Baseline** (lr=2e-4, wd=5e-5, T=160) | 13.2 | 7.9 | 30.8 | 6.5 | 0.2626 | 369n5uv8 |
| 1 | lr=1.5e-4 | **12.6** | 7.9 | 30.9 | 6.8 | 0.2576 | i8xwj9zn |
| 2 | lr=2.5e-4 | 13.3 | 7.8 | **30.3** | 6.5 | 0.2636 | 2bht89ay |
| 3 | T_max=140 | 12.9 | **7.7** | 31.5 | 6.5 | 0.2610 | wrbmgygu |
| 4 | T_max=180 | 13.6 | 7.9 | 31.0 | **6.3** | 0.2730 | qmsobmzv |
| 5 | wd=1e-5 | 13.0 | 8.0 | 31.0 | **6.3** | 0.2675 | cofd0trk |
| 6 | wd=1e-4 | 13.0 | 7.9 | 30.9 | **6.3** | 0.2645 | 80aq4s2f |
| 7 | ema=0.9995 | 13.1 | 8.0 | 31.0 | **6.3** | 0.2695 | o4dh75fa |
| **Mean** | — | **13.1** | **7.9** | **30.9** | **6.4** | **0.2649** | — |

### Ensemble Evaluation

| Ensemble | p_in | p_oodc | p_tan | p_re |
|----------|------|--------|-------|------|
| **Diverse 8-model** (all configs) | **12.3** | **7.0** | **29.8** | **5.9** |
| Diverse 4-model (best subset) | 12.2 | 6.9 | 29.9 | 6.1 |
| **Same-config 8-seed baseline** | **12.4** | **6.7** | **29.4** | **5.8** |

### Diverse-8 vs Same-Config-8 Baseline

| Metric | Diverse-8 | Same-Config-8 | Delta |
|--------|-----------|---------------|-------|
| p_in | **12.3** | 12.4 | **-0.8%** |
| p_oodc | 7.0 | **6.7** | +4.5% |
| p_tan | 29.8 | **29.4** | +1.4% |
| p_re | 5.9 | **5.8** | +1.7% |

### Commands used

```bash
# Training (8 configs, all seed 42)
# GPU 0: baseline control
CUDA_VISIBLE_DEVICES=0 python train.py --agent nezuko --wandb_name "nezuko/baseline-lr2e4-s42" --wandb_group "phase6/diverse-hparam-ensemble" --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --weight_decay 5e-5 --cosine_T_max 160 --seed 42 [+ all baseline flags]
# GPU 1-7: varied lr, T_max, wd, ema_decay (see PR instructions)

# Ensemble evaluation
python eval_ensemble.py --run_ids 369n5uv8 i8xwj9zn 2bht89ay wrbmgygu qmsobmzv cofd0trk 80aq4s2f o4dh75fa --asinh_scale 0.75
```

- **W&B group:** `phase6/diverse-hparam-ensemble`
- **Peak VRAM:** ~38 GB per model
- **Training time:** ~180 min each (hit wall-clock limit)

### What happened

The diverse hyperparameter ensemble produces **mixed results** compared to the same-config 8-seed baseline:

- **p_in improved slightly** (12.3 vs 12.4, -0.8%) — the diversity helps here
- **p_oodc, p_tan, p_re all slightly worse** (+1.4% to +4.5%)

The **same-config seed diversity** actually outperforms hyperparameter diversity for the OOD and tandem splits. This is likely because:

1. **All models share the same seed (42)** — so initialization is identical, limiting the diversity benefit to hyperparameter differences alone
2. **The hyperparameter differences are relatively small** (e.g., lr 1.5e-4 vs 2.5e-4) — they don't produce wildly different models
3. **Some configs are individually weaker** (e.g., T_max=180 has p_in=13.6, lr=1.5e-4 has p_re=6.8) — weak models drag down the ensemble

Interestingly, the **best 4-of-8 diverse subset** (12.2, 6.9, 29.9, 6.1) outperforms the full 8 on p_in and p_oodc, suggesting that trimming the weakest models helps.

### Suggested follow-ups

- **Mixed ensemble:** Combine diverse-hparam models WITH different seeds (cross-product of seed × hparam diversity) — this should get the best of both worlds
- **Weighted ensemble:** Use validation performance to weight model contributions rather than simple averaging
- The 4-of-8 subset being competitive suggests diminishing returns from 8 models — investigate optimal ensemble size

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2091 because ADVISOR: Excellent analysis. Key findings:

1. **Diverse-hparam ensemble DOES NOT beat same-config seed ensemble.** p_oodc +4.5%, p_tan +1.4%, p_re +1.7% all WORSE. Only p_in marginally better (-0.8%).

2. **Seed diversity > hyperparameter diversity** for this architecture. Same-config models with different seeds make different random errors (good for averaging). Different hyperparameters make systematic tradeoffs (some configs weaker on specific metrics, dragging ensemble down).

3. The best-4-of-8 subset (p_in=12.2) is interesting but relies on model selection, which askeladd (#2089) showed provides only marginal and inconsistent gains.

4. Note: the diverse ensemble was compared against the old PR #2076 baseline (seeds 42-49, p_in=12.4). The current baseline (PR #2080, seeds 66-73) is even stronger: p_in=12.2, p_oodc=6.7, p_tan=29.1, p_re=5.8. Against this, the diverse ensemble loses on ALL metrics.

**Conclusion:** Hyperparameter diversity adds tradeoffs, not complementary error reduction. Future ensemble improvement needs qualitatively different methods (different losses, architectures), not just different hyperparameters. This aligns with askeladd's finding that all same-architecture models deserve equal weight.

Closing and assigning a new experiment. Good work — negative results that tell us WHERE NOT to look are valuable..
